### PR TITLE
ESIMW-1249: Java 11 support

### PR DESCRIPTION
**Pull request by jwbennet**
This modifies the build process and several dependencies so that Kuali Rice will build and run against Java 11.  This includes a number of changes:

* [Backport some changes made to make Rice 2.5 work on Java 8](https://github.com/kuali/rice/commit/482b0390117f7536b7124a578f75cb67f09325bc).  The primary change here is to switch the Groovy compiler and Eclipselink static weaving plugins to different versions which make them work better on newer versions of Java.  Some of the dependencies used here were also updated to newer versions which made them work for Java 11.
* Updated Mockito and Powermock to newer versions which support Java 11.  This required fixing a number of tests due to some behavior changes in the new version of Mockito.  A good majority of the changes to test classes were related to this.
* Fixed a number of tests which started failing because results would come back in slightly different orders.
* Added some missing JPA `@Convert` attributes for boolean properties which are stored as 1 or 0.
* Removed the `KRADLibraryPropertiesDoclet` class which was used for something optional in the build process.  The support for doclets was removed from the JDK and I wasn't seeing a quick dependency to bring in to allow this to continue to work so I removed it instead.
* Removed the use of the `cxf-java2ws-plugin`.  This is used to generate WSDLs from our Java classes so they can be distributed through Maven central and used for version compatibility tests.  We don't run those tests and this plugin does not appear to work on Java 11.  It is possible this could be added back in with a major CXF version update, but I thought it would be easier to update if we didn't do that major version change.
* Removed some plugin management configuration for the `org.eclipse.m2e:lifecycle-mapping` plugin.  It is only needed for Eclipse and should be completely removed, but for now I have only removed the mappings which referenced plugins I had removed to avoid build issues.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -148,20 +148,21 @@
     <closure.version>v20130603</closure.version>
     <wro4j.version>1.7.0</wro4j.version>
     <glassfish-javax-json.version>1.0.1</glassfish-javax-json.version>
-    <groovy.version>1.8.5</groovy.version>
+    <groovy.version>2.4.15</groovy.version>
     <guava.version>11.0</guava.version>
-    <eclipselink.version>2.5.1</eclipselink.version>
+    <eclipselink.version>2.7.4-SNAPSHOT</eclipselink.version>
```
Comment (jwbennet) [inline pom.xml:143]: This pre-release version is necessary for getting the static weaving to work with Java 11.

File: pom.xml [hunk 2]
```diff
@@ -211,7 +212,7 @@
     <soa.model.core.version>1.2.2</soa.model.core.version>
     <httpclient.version>4.2.3</httpclient.version>
     <spring-modules-ojb.version>0.8a</spring-modules-ojb.version>
-    <spring.version>3.2.3.RELEASE</spring.version>
+    <spring.version>3.2.18.RELEASE</spring.version>
```
Comment (jwbennet) [inline pom.xml:205]: This updates to the newest version of Spring 3 and things seem to generally work with it.  We could consider also upgrading to Spring 4 by backporting some changes from KualiCo's version of Rice, but I'm thinking that should be a separate PR if we decide to pursue that.

File: pom.xml [hunk 3]
```diff
@@ -181,8 +182,8 @@
     <kuali-mvn.version>1.0.2</kuali-mvn.version>
     <log4j.version>1.2.16</log4j.version>
     <mail.version>1.4.4</mail.version>
-    <mockito.version>1.10.8</mockito.version>
-    <powermock.version>1.6.1</powermock.version>
+    <mockito.version>2.23.0</mockito.version>
+    <powermock.version>2.0.0-RC.4</powermock.version>
```
Comment (jwbennet) [inline pom.xml:176]: The full release of this version of Powermock is still not available so bringing in the most recent release candidate.

File: pom.xml [hunk 4]
```diff
@@ -860,60 +861,6 @@
                       <exclude>org.ow2.spec.ee:ow2-jta-1.1-spec</exclude>
                       <exclude>jta</exclude>
 
-                      <!-- activation apis - in jdk 6 -->
-                      <exclude>javax.activation</exclude>
```
Comment (jwbennet) [inline pom.xml:864]: This block of removals are part of the maven enforcer plugin's configuration.  In the past this made sense because all of these were delivered with the JDK so they should have been excluded.  They were all removed in Java 11 and need to be brought in as external dependencies now so this had to be removed to allow for them to be brought in.

File: pom.xml [hunk 5]
```diff
@@ -1122,6 +1043,12 @@
               <artifactId>jaxb-impl</artifactId>
               <version>${jaxb-impl.version}</version>
             </dependency>
+            <dependency>
+              <groupId>com.sun.xml.ws</groupId>
```
Comment (jwbennet) [inline pom.xml:1037]: This dependency seems to bring in all of the JAX-WS and JAXB dependencies Rice needs which were removed in Java 11.

File: pom.xml [hunk 6]
```diff
@@ -1267,11 +1194,9 @@
                 </pluginExecution>
                 <pluginExecution>
                 	<pluginExecutionFilter>
-                		<groupId>au.com.alderaan</groupId>
-                		<artifactId>
-                			eclipselink-staticweave-maven-plugin
-                		</artifactId>
-                		<versionRange>[1.0.4,)</versionRange>
+                      <groupId>de.empulse.eclipselink</groupId>
+                      <artifactId>staticweave-maven-plugin</artifactId>
```
Comment (jwbennet) [inline pom.xml:1188]: This plugin was changed to a different version which supports newer versions of Eclipselink.

File: pom.xml [hunk 7]
```diff
@@ -1508,18 +1439,37 @@
...
+          <url>http://nexus.kuali.org/content/groups/public</url>
+          <releases>
+              <enabled>true</enabled>
+          </releases>
+          <snapshots>
+              <enabled>false</enabled>
+          </snapshots>
+      </repository>
+      <repository>
+          <id>oss.sonatype.org</id>
```
Comment (jwbennet) [inline pom.xml:1444]: This additional repository was required to bring in the pre-release version of Eclipselink

File: pom.xml [hunk 8]
```diff
@@ -1508,18 +1439,37 @@
...
+          <name>OSS Sonatype Staging</name>
+          <url>https://oss.sonatype.org/content/groups/staging</url>
+      </repository>
+  </repositories>
+
+  <pluginRepositories>
+    <pluginRepository>
+      <id>bintray</id>
+      <name>Groovy Bintray</name>
+      <url>https://dl.bintray.com/groovy/maven</url>
```
Comment (jwbennet) [inline pom.xml:1454]: This additional plugin repository was required for bringing in the latest versions of some Groovy plugins

File: rice-middleware/web/src/main/java/javax/rmi/PortableRemoteObject.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+import java.rmi.NoSuchObjectException;
+import java.rmi.Remote;
+import java.rmi.RemoteException;
+
+/**
+ * This is being provided as a no-op shim since JOTM depends on this class but
+ * it does nothing with it.  This class was removed from the JDK for Java 11
+ * when CORBA was removed.
+ */
```
Comment (jwbennet) [inline rice-middleware/web/src/main/java/javax/rmi/PortableRemoteObject.java:11]: This comment mostly explains what this class is used for.  It is provided because I wasn't finding a replacement JAR which satisfied JOTM's needs.

File: rice-framework/krad-data/src/main/java/org/kuali/rice/krad/data/jpa/eclipselink/EclipseLinkJpaMetadataProviderImpl.java [hunk 10]
```diff
@@ -265,30 +265,35 @@ protected void populateImplementationSpecificRelationshipLevelMetadata(DataObjec
...
+						if (parentFieldName != null) {
+							attributeRelationships
+									.add(new DataObjectAttributeRelationshipImpl(parentFieldName, childFieldName));
+							break;
+						} else {
+							LOG.warn("Unable to find parent field reference.  There may be a JPA mapping problem on " +
+									referencedEntityType.getJavaType() + ": " + relationship);
+						}
+					}
+				}
```
Comment (athill) [inline rice-framework/krad-data/src/main/java/org/kuali/rice/krad/data/jpa/eclipselink/EclipseLinkJpaMetadataProviderImpl.java:291]: I'd like a little more information on this changed block. Including why the special handling of child fields was able to be removed.

File: pom.xml [hunk 11]
```diff
@@ -122,8 +122,8 @@
     <commons-pool.version>1.5.7</commons-pool.version>
     <commons-digester.version>2.0</commons-digester.version>
     <commons-validator.version>1.3.1</commons-validator.version>
-    <cxf.version>2.7.0</cxf.version>
-    <jaxb-impl.version>2.2.6</jaxb-impl.version>
+    <cxf.version>2.7.17</cxf.version>
+    <jaxb-impl.version>2.3.1</jaxb-impl.version>
     <!--
```
Comment (burnumd) [inline pom.xml]: Can this comment be removed now that we're moving past jaxb-impl 2.2.x entirely?

File: rice-framework/krad-data/src/test/java/org/kuali/rice/krad/data/jpa/JpaMetadataProviderTest.java [hunk 12]
```diff
@@ -153,12 +143,12 @@ public void testGetMetadataForCollection() {
 		assertNotNull("Collections object should not be null", collections);
 		assertEquals("Collections size incorrect", 4, collections.size());
 
-		DataObjectCollection collection = collections.get(0);
+		DataObjectCollection collection = metadata.getCollection("collectionProperty");
 		assertEquals("property name incorrect", "collectionProperty", collection.getName());
```
Comment (athill) [inline rice-framework/krad-data/src/test/java/org/kuali/rice/krad/data/jpa/JpaMetadataProviderTest.java:147]: I like more explicit `metadata.getCollection("collectionProperty")`, though it kind of makes me want a variable for "collectionProperty", as it's now used twice.

File: rice-framework/krad-data/src/main/java/org/kuali/rice/krad/data/jpa/eclipselink/EclipseLinkJpaMetadataProviderImpl.java [hunk 13]
```diff
@@ -241,7 +241,7 @@ protected String getPropertyNameFromDatabaseColumnName(ManagedType entityType, S
      */
 	@Override
 	protected void populateImplementationSpecificRelationshipLevelMetadata(DataObjectRelationshipImpl relationship,
-			SingularAttribute<?, ?> rd) {
+																		   SingularAttribute<?, ?> rd) {
```
Comment (burnumd) [inline rice-framework/krad-data/src/main/java/org/kuali/rice/krad/data/jpa/eclipselink/EclipseLinkJpaMetadataProviderImpl.java:244]: Looks like all the whitespace in this file got replaced with tabs. I'm having trouble sussing out what was changed in the logic. Are the PK attributes now just extracted outside the for loop or is something else different happening here?

File: rice-middleware/impl/src/main/java/org/kuali/rice/kim/bo/ui/KimDocumentBoActivatableToFromBase.java [hunk 14]
```diff
@@ -15,14 +15,17 @@
...
 import javax.persistence.Column;
+import javax.persistence.Convert;
 import javax.persistence.MappedSuperclass;
 import javax.persistence.Transient;
 import java.sql.Timestamp;
 
 /**
- * This is a description of what this class does - shyu don't forget to fill this in. 
- * 
+ * This is a description of what this class does - shyu don't forget to fill this in.
```
Comment (burnumd) [inline rice-middleware/impl/src/main/java/org/kuali/rice/kim/bo/ui/KimDocumentBoActivatableToFromBase.java:27]: 😬

File: rice-middleware/impl/src/test/java/org/kuali/rice/kew/rule/GenericAttributeContentTest.java [hunk 15]
```diff
@@ -36,22 +36,22 @@
...
-                                                  "        <name>car</name>" +
-                                                  "        <value>KIT</value>" +
-                                                  "      </field>" +
                                                   "      <field>" +
                                                   "        <name>driver</name>" +
                                                   "        <value>hasselhof</value>" +
                                                   "      </field>" +
+                                                  "      <field>" +
+                                                  "        <name>car</name>" +
+                                                  "        <value>KIT</value>" +
```
Comment (burnumd) [inline rice-middleware/impl/src/test/java/org/kuali/rice/kew/rule/GenericAttributeContentTest.java:45]: Not saying it needs to be fixed since there are probably tests that rely on it being this way, but that is not the name of [the car in Knight Rider](https://en.wikipedia.org/wiki/KITT).

**Review by burnumd — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 11 above)
Comment (jwbennet) [reply pom.xml]: Not necessarily.  The plugin this comment refers to was removed since it was still having issues so I think this may be relevant if we ever try to add it back in.  I was trying to only modify things that I had to while going through this.

**Review by athill — CHANGES_REQUESTED** [review]
Pretty huge reduction in configuration :)

File: rice-framework/krad-data/src/main/java/org/kuali/rice/krad/data/jpa/eclipselink/EclipseLinkJpaMetadataProviderImpl.java (hunk 13 above)
Comment (jwbennet) [reply rice-framework/krad-data/src/main/java/org/kuali/rice/krad/data/jpa/eclipselink/EclipseLinkJpaMetadataProviderImpl.java:244]: This was part of the first bullet point in the PR's description.  This was backported from Rice 2.5 verbatim.

File: rice-middleware/impl/src/main/java/org/kuali/rice/kim/bo/ui/KimDocumentBoActivatableToFromBase.java (hunk 14 above)
Comment (jwbennet) [reply rice-middleware/impl/src/main/java/org/kuali/rice/kim/bo/ui/KimDocumentBoActivatableToFromBase.java:27]: Yeah, this kind of thing is all over the place.  You may find more success ignoring whitespace changes when reviewing the PR since IntelliJ changed whitespace on a bunch of lines I didn't actually modify while I was working through things.

File: pom.xml (hunk 11 above)
Comment (burnumd) [reply pom.xml]: Ah, ok. You might consider updating it to make that clear because it reads to me like it's referring to the property below.

File: rice-framework/krad-data/src/main/java/org/kuali/rice/krad/data/jpa/eclipselink/EclipseLinkJpaMetadataProviderImpl.java (hunk 10 above)
Comment (jwbennet) [reply rice-framework/krad-data/src/main/java/org/kuali/rice/krad/data/jpa/eclipselink/EclipseLinkJpaMetadataProviderImpl.java:291]: I'll refer you to the first bullet point in the PR's description.

File: rice-framework/krad-data/src/test/java/org/kuali/rice/krad/data/jpa/JpaMetadataProviderTest.java (hunk 12 above)
Comment (jwbennet) [reply rice-framework/krad-data/src/test/java/org/kuali/rice/krad/data/jpa/JpaMetadataProviderTest.java:147]: This relates to the bullet in the PR's description where there were some ordering issues in the tests where things come back in a different order.  I'm also not inclined to clean up these tests more than necessary because if I started fixing issues like making "collectionProperty" a variable I'd never be able to stop ;).

File: pom.xml (hunk 11 above)
Comment (jwbennet) [reply pom.xml]: It is referring to the property below.  I'm also not inclined to change a bunch of code which I didn't modify here in case it has some other side effect.

**Review by athill — APPROVED** [review]

File: rice-framework/krad-data/src/main/java/org/kuali/rice/krad/data/jpa/eclipselink/EclipseLinkJpaMetadataProviderImpl.java (hunk 13 above)
Comment (burnumd) [reply rice-framework/krad-data/src/main/java/org/kuali/rice/krad/data/jpa/eclipselink/EclipseLinkJpaMetadataProviderImpl.java:244]: Ah, ok, thanks. I wasn't clear on its relevance to the rest of the changes.

File: pom.xml (hunk 11 above)
Comment (burnumd) [reply pom.xml]: Are you talking about updating the WorkflowDocumentService? I'm just saying this comment no longer makes sense as written since it refers to us downgrading the plugin when we're now on an even newer version. I still think it would be helpful to have an update in the comment that explains what you mentioned above: that the problem causing plugin has been removed and that the comment remains for historical purposes.

Comment (jwbennet) [reply pom.xml]: Turns out that property was unused so I removed it.

**Review by burnumd — APPROVED** [review]

File: pom.xml [hunk 16]
```diff
@@ -122,8 +122,8 @@
     <commons-pool.version>1.5.7</commons-pool.version>
     <commons-digester.version>2.0</commons-digester.version>
     <commons-validator.version>1.3.1</commons-validator.version>
-    <cxf.version>2.7.0</cxf.version>
-    <jaxb-impl.version>2.2.6</jaxb-impl.version>
+    <cxf.version>2.7.17</cxf.version>
```
Comment (ewestfal) [inline pom.xml:125]: You'll definately want to test the SOAP web services on jdk11. I recall there being some weirdness with how cxf wrapped the XML envelope when using the xerces implementation that was part of the jdk vs the one that was standalone and I have no idea how/if that will work with jdk11.

Also to consider, if we have issues with the web services, we may consider taking the plunge and moving to CXF 3.2.x. The 2.7.x version is no longer supported by CXF so may have security vulnerabilities either reported or unreported. Rice 2.6 is using CXF 3.2.6 so there is at least a functioning example out there somewhere.

File: pom.xml (hunk 2 above)
Comment (ewestfal) [reply pom.xml:205]: Yeah, we'd have to go through the cxf upgrade (and a number of others) that I mentioned above in order to do that. Probably worthwhile seeing if we can get it to work at this stage. We may still ultimately want to do the upgrade so that we can get on a supported version of spring and other dependencies to improve our security profile on this application since it will probably be hanging around for awhile still.

File: rice-middleware/impl/src/main/java/org/kuali/rice/kim/bo/ui/KimDocumentBoActivatableToFromBase.java (hunk 14 above)
Comment (ewestfal) [reply rice-middleware/impl/src/main/java/org/kuali/rice/kim/bo/ui/KimDocumentBoActivatableToFromBase.java:27]: This is what code looks like that *didn't* go through a PR/code review ;)

File: rice-middleware/impl/src/test/java/org/kuali/rice/kew/rule/GenericAttributeContentTest.java (hunk 15 above)
Comment (ewestfal) [reply rice-middleware/impl/src/test/java/org/kuali/rice/kew/rule/GenericAttributeContentTest.java:45]: Don't hassle the Hoff

File: rice-middleware/ksb/client-impl/pom.xml [hunk 17]
```diff
@@ -174,7 +174,7 @@
     </dependency>
     <!-- ASM as an explicit dependency, see KULRICE-6353 -->
     <dependency>
-      <groupId>asm</groupId>
+      <groupId>org.ow2.asm</groupId>
```
Comment (ewestfal) [inline rice-middleware/ksb/client-impl/pom.xml:177]: One thing just in general to check on these, but it would probably be a good idea build the WAR for Rice standalone from these updates, then extract it and check WEB-INF/lib and make sure it doesn't end up with both of the asm jars in it in case that's coming in some other way. In general, I would recommend building the war and just looking at the jars, maybe compare it with the prior war file just to make sure there is nothing weird and no duplicates with different versions.

Also, another kind of weird dependency that you might check out is cglib-nodep. If that ends up in the war for some reason, it has ASM dependencies built into it (and repackaged so as not to conflict). But if ASM 7 is what is required for jdk11 then you might want to check that our cglib is properly up to date too.

**Review by ewestfal — APPROVED** [review]
Generally speaking, from what I can tell this looks good. Lots of testing needed of course.

As I mention in some of the comments, we may want to consider making the move to newer versions of Spring and CXF. But getting this to work is a start and we can take it from there.

File: pom.xml (hunk 16 above)
Comment (jwbennet) [reply pom.xml:125]: I just tried this and I think I'm hitting the error you're describing here.  Does this seem familar?

```
java.lang.ClassCastException: com.sun.org.apache.xerces.internal.dom.ElementNSImpl cannot be cast to javax.xml.soap.SOAPElement
```

Comment (jwbennet) [reply pom.xml:125]: This is still not working but I think I'm going to try updating to Spring 4 and CXF 3 in a separate PR to see if I can get this resolved.

File: rice-middleware/ksb/client-impl/pom.xml (hunk 17 above)
Comment (jwbennet) [reply rice-middleware/ksb/client-impl/pom.xml:177]: I updated cglib-nodep to its latest version as well
